lib/Doctrine/ORM/Internal/Hydration/SimpleObjectHydrator.php

```diff
@@ -106,7 +106,11 @@ protected function hydrateRowData(array $sqlResult, array &$result)
 
             $entityName = $discrMap[$sqlResult[$discrColumnName]];
 
+            $discriminator = $sqlResult[$discrColumnName];
+
             unset($sqlResult[$discrColumnName]);
+
+            $this->removeColumnsNotRelatedToEntityDiscriminator($discriminator, $sqlResult);
```

Comment:
When `hydrateRowData` is called, the resultset mappings are already known, so we should be able to intersect the row with the resultset columns upfront.